File: 동적계획법/leetcode724(Easy)-Find Pivot Index.js
// 50ms(99%), 45MB(20%)

/**
 * @param {number[]} nums
 * @return {number}
 */
var pivotIndex = function (nums) {
  const n = nums.length
  const sums = []
  let acc = 0
  nums.forEach((num) => {
    acc += num
    sums.push(acc)
  })

  const maxSum = sums[n - 1]
  for (let i = 0; i < n; i += 1) {
    const left = sums[i] - nums[i]
    const right = maxSum - sums[i]
    if (left === right) {
      return i
    }
  }

  return -1
}

File: 동적계획법/leetcode724(Easy)-Find Pivot Index.py
"""
145ms(91%), 15.2MB(30%)
"""
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        sums = []
        acc = 0
        for num in nums:
            acc += num
            sums.append(acc)

        maxSum = sums[-1]
        for i in range(len(nums)):
            left = sums[i] - nums[i]
            right = maxSum - sums[i]
            if left == right:
                return i

        return -1